common/src/utils.rs

```diff
@@ -1,49 +1,153 @@
 #[cfg(not(feature = "std"))]
 use alloc::vec::Vec;
+use core::ops::Deref;
 #[cfg(feature = "std")]
 use std::vec::Vec;
 
 use crate::circuit::F;
-use plonky2::field::types::{Field, PrimeField64};
+use anyhow::anyhow;
+use plonky2::field::types::{Field, Field64, PrimeField64};
 use plonky2::hash::hash_types::HashOut;
 
-pub const BYTES_PER_ELEMENT: usize = 8;
-pub const FELTS_PER_U128: usize = 2;
+pub const INJECTIVE_BYTES_PER_ELEMENT: usize = 4;
+pub const DIGEST_BYTES_PER_ELEMENT: usize = 8;
+pub const FELTS_PER_U128: usize = 4;
+pub const FELTS_PER_U64: usize = 2;
 pub const DIGEST_NUM_FIELD_ELEMENTS: usize = 4;
 
-pub const ZERO_DIGEST: Digest = [F::ZERO; 4];
+pub const ZERO_DIGEST: Digest = [F::ZERO; DIGEST_NUM_FIELD_ELEMENTS];
+pub const BIT_32_LIMB_MASK: u64 = 0xFFFF_FFFF;
 
-pub type Digest = [F; 4];
+pub type Digest = [F; DIGEST_NUM_FIELD_ELEMENTS];
 pub type PrivateKey = [F; 4];
 
+#[derive(Default, Debug, Clone, Copy, PartialEq, Eq)]
+pub struct BytesDigest(pub [u8; 32]);
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum DigestError {
+    ChunkOutOfFieldRange { chunk_index: usize, value: u64 },
+    InvalidLength { expected: usize, got: usize },
+}
+
+impl TryFrom<&[u8]> for BytesDigest {
+    type Error = DigestError;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        let bytes: [u8; 32] = value.try_into().map_err(|_| DigestError::InvalidLength {
+            expected: 32,
+            got: value.len(),
+        })?;
+        for (i, chunk) in bytes.chunks(8).enumerate() {
+            let v = u64::from_le_bytes(chunk.try_into().unwrap());
+            if v >= F::ORDER {
```

Comment:
this protects us against accidentally putting a non-poseidon digest into the try_from, but doesn't enforce that check in the circuit. Do we need to enforce it in the circuit?